File: Kimyerim1935/프로그래머스_150369.js
/*
 * Programers
 * https://school.programmers.co.kr/learn/courses/30/lessons/150369
 * Problem Number: 150369
 */


function solution(cap, n, deliveries, pickups) {
    let answer = 0;
    let delSum = deliveries.reduce((a,b)=>a+b,0);
    let pickSum = pickups.reduce((a,b)=>a+b,0);
//배달해야하는 화물, 수거해야할 화물 모두 0이되면 종료
    while(delSum !== 0 || pickSum !== 0){
        deleteZero(deliveries);
        deleteZero(pickups);
        let len = Math.max(deliveries.length, pickups.length);
        answer += len*2;
        delSum -= delItem(deliveries, cap);
        pickSum -= delItem(pickups, cap);
    }
    return answer;
}
//뒤에서 부터 0이 있으면 제거해줌
const deleteZero = (arr)=>{
    for(let i=arr.length-1; i>=0; i--){
        if(arr[i] === 0) arr.pop();
        else return;
    }
}
//현재 cap에 맞게, 뒤에서부터 빼줌
//ex) 0 3 2에 cap이 3이면 => 0 2 0
const delItem = (arr,cap) =>{
    let cnt = 0;
    for(let i=arr.length-1; i>=0; i--){
        if(arr[i] >= cap){
            arr[i] -= cap;
            cnt += cap;
            break;
        } 
        else{
            cap -= arr[i];
            cnt += arr[i];
            arr[i] = 0;
        } 
    }
    return cnt;
}

File: soonba/프로그래머스_150369.java

/*
 * https://school.programmers.co.kr/learn/courses/30/lessons/150369
 * Problem Number: 150369
 */

/*
key point >> 최소로 움직이려면, 제일 끝에서부터 배송과 회수를 해야함
delivery, pickup 변수는 각 집에 배송, 회수하고 남은 양을 저장(다음 집에서 사용)
각 집을 for 문으로 끝에서부터 순회하며, delivery 와 pickup 변수가 해당 집의 배송과 회수 각각보다 클 때까지 cap을 더한다.
해당 집에 필요한 방문은 끝났으므로 해당 집의 delivery 와 pickup 을 빼고 다음 집 순회
결과를 출력한다.
*/
public class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {

        int delivery = 0, pickup = 0;
        long moveCount = 0;
        for (int i = n-1; i > -1 ; i--) {
            int cnt = 0;
            while(deliveries[i] > delivery || pickups[i] > pickup) {
                delivery += cap;
                pickup += cap;
                cnt++;
            }
            moveCount += 2L * cnt * (i+1);
            delivery -= deliveries[i];
            pickup -= pickups[i];
        }

        return moveCount;
    }
}


